trusted-types/block-string-assignment-to-Element-setAttributeNode.html

```diff
@@ -0,0 +1,79 @@
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="utf-8">
+<meta http-equiv="Content-Security-Policy" content="require-trusted-types-for 'script';">
+<title>trusted-types (TT): setAttributeNode with node from non-TT realm</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+</head>
+<body>
+<div id="nonSVGTestElements">
+    <iframe srcdoc="v"></iframe>
+    <embed src="v">
+    <script src="v"></script>
+    <object data="v"></object>
+    <object codebase="v"></object>
+</div>
+<svg id="svgTestElements">
+    <script href="v"></script>
+    <script xlink:href="v"></script>
+</svg>
+<script>
+    const passThroughPolicy = trustedTypes.createPolicy("passThrough", { createHTML: s => s });
+
+    function runTest(aTestElement) {
+        const testAttr = aTestElement.attributes[0];
+
+        async_test(t => {
+            const sourceFrame = document.createElement("iframe");
+
+            // The markup requires the parent element to ensure the attribute is associated with the
+            // correct namespace.
+            sourceFrame.srcdoc = passThroughPolicy.createHTML(
+            `<!DOCTYPE html>
+            <head>
+            <meta charset="utf-8">
+            </head>
+            <body>
+                <` + aTestElement.parentElement.localName + `>
+                    <` + aTestElement.localName + ` ` + testAttr.name + `="` + testAttr.value + `">
+                    </` + aTestElement.localName + `>
+                </` + aTestElement.parentElement.localName + `>
+            doc without TT CSP.
+            </body>`);
+
+            t.add_cleanup(() => {
+                sourceFrame.remove();
+            });
+
+            sourceFrame.addEventListener("load", t.step_func_done(() => {
+                // A window is a global object which has 1-to-1 mapping to a realm, see the first
+                // note of <https://html.spec.whatwg.org/#realms-settings-objects-global-objects>
+                // and its following paragraph. Here, `aTestElement`'s and `sourceAttr`'s windows
+                // differ, hence they belong to different realms.
+
+                const sourceElement = sourceFrame.contentDocument.body.querySelector(aTestElement.localName);
+                const sourceAttr = sourceElement.getAttributeNode(testAttr.name);
+                sourceElement.removeAttributeNode(sourceAttr);
+
+                assert_throws_js(TypeError, () => { aTestElement.setAttributeNode(sourceAttr); });
```

Comment:
Based on the requirements it seems that setting an attribute from the same document would end up throwing here as well, right?

Also, why is this called "block-string-assignment-*" when it's testing a cross-global thing?

Reply:
> Based on the requirements it seems that setting an attribute from the same document would end up throwing here as well, right?

Yes. The aspect which is tested here, is that element's node document's global is taken in step 5 of https://w3c.github.io/trusted-types/dist/spec/#validate-attribute-mutation. The `sourceAttr`'s global's CSP, i.e. the iframe's, doesn't enforce TT.
> 
> Also, why is this called "block-string-assignment-*" when it's testing a cross-global thing?

Because the attribute's value is a string. Would `block-cross-global-assignment-to-Element-setAttributeNode` be clearer?

Reply:
That algorithm only cares about element's node document's global though. So a relevant test would have to move the element between documents?

Reply:
> That algorithm only cares about element's node document's global though. So a relevant test would have to move the element between documents?

To over-communicate: setting an attribute node from the same document would throw too, but that would be outside of the threat-model because the web-app author is considered to be non-malicious according to https://w3c.github.io/trusted-types/dist/spec/#security-considerations. Hence, it requires no test.

However, when the attribute node stems from a different document, e.g. an iframe, controlled by an attacker, that's inside the threat-model and requires the current test.

Reply:
Moving elements between documents, e.g. adopting an iframe's `<script>` element, seems a security issue too. But non to be addressed by Trusted Types.

Reply:
Unless I missed something I'll add a comment to the test-code about the threat-model.

Reply:
So this test isn't about testing that a certain global is as mentioned in https://github.com/web-platform-tests/wpt/pull/44323#discussion_r1540871059.

Reply:
But the global confusion is why this test was needed in the first place. To ensure that the policy is picked up from the correct global.

That's https://github.com/w3c/trusted-types/pull/418#discussion_r1463486461 which is linked from https://github.com/w3c/trusted-types/issues/425.

Reply:
Agree. So a minimal example for that scenario is adopting an iframe which was nested in another iframe and calling `setAttribute("srcdoc", "v")` (or `setAttributeNode`) for the adopted iframe: https://jsfiddle.net/ta01od8m/1/. Correct?

Reply:
Yeah, that looks correct.